packages/api/lambdas/bulk-operation.js

```diff
@@ -118,26 +128,67 @@ function applyWorkflowToGranules({
 async function bulkGranuleDelete(payload) {
   const granuleIds = await getGranuleIdsForPayload(payload);
   const granuleModel = new GranuleModel();
+  const granulePgModel = new GranulePgModel();
   const forceRemoveFromCmr = payload.forceRemoveFromCmr === true;
+  const knex = await getKnexClient({ env: process.env });
   const deletedGranules = [];
   await pMap(
     granuleIds,
     async (granuleId) => {
+      const collectionPgModel = new CollectionPgModel();
+
+      let dynamoGranule;
+      let pgGranule;
+
+      // Try to get the Dynamo record. If it cannot be found, just log
+      // the error and skip it.
       try {
-        const granule = await granuleModel.getRecord({ granuleId });
-        if (granule.published && forceRemoveFromCmr) {
-          await granuleModel.unpublishAndDeleteGranule(granule);
-        } else {
-          await granuleModel.delete(granule);
-        }
-        deletedGranules.push(granuleId);
+        dynamoGranule = await granuleModel.getRecord({ granuleId });
       } catch (error) {
         if (error instanceof RecordDoesNotExist) {
           log.info(`Granule ${granuleId} does not exist or was already deleted, continuing`);
           return;
         }
         throw error;
       }
+
+      // Try to get the PG record. If it cannot be found, ignore it and
+      // move along to unpublishing and deleting only the Dynamo record.
+      // If another error is thrown, throw it here.
+      try {
+        const collectionCumulusId = await collectionPgModel.getRecordCumulusId(
+          knex,
+          deconstructCollectionId(dynamoGranule.collectionId)
+        );
+
+        pgGranule = await granulePgModel.get(knex,
+          {
+            granule_id: granuleId,
+            collection_cumulus_id: collectionCumulusId,
+          });
+      } catch (error) {
+        if (!(error instanceof RecordDoesNotExist)) {
+          throw error;
+        }
+      }
```

Comment:
Yeah. I know we don't really want to get these records before deleting. Totally open to refactoring suggestions here. This was a compromise to try and avoid refactoring a ton of stuff from the previous work.

Reply:
This goes back a bit to the arch meeting, but yes, the only way to do this other than a direct cumulus_id reference would be to do the work to do a delete with the unique constrained keys, which we do have here........

The functional problem I see with this approach is: 

1)  Getting the Dynamo Record
2) Pulling the collection/granule cumulus_id based on that 
3) Optionally - Unpublish (getting the updated dynamo record and/or PG record)
4) Delete Granules and Files (using the value from 3 if present *else* 2

We should reorder this such that it doesn't bother doing the get queries if unpublish is going to happen.

Reply:
Totally makes sense. See what you think of this:

https://github.com/nasa/cumulus/commit/4b006bd0b60a9c8fc412c8e66795ea5a61184280

The idea here is to get the Dynamo Granule and check if it's published right away. If it is and we're doing a "forceRemoveFromCMR", we'll do the update on both Dynamo and PG records (if we have them). Then we can use the response from the update if it went through. 

This should bail earlier if a granule is published and save a PG fetch request if the granule was updated.

Reply:
Also note the new private function `_getPgGranuleByCollection`. This lives in the bulk-operation Lambda but I hemmed and hawed a bit over where to put it. The other option is to drop it in /lib somewhere, likely either granules.js or a new file. But then I don't know if I (or anyone else) would know to look for it there if they needed it. I also don't know if this will ever be used again outside of this Lambda.

Reply:
Nice work!   I'm ok with YAGNI-ing the concern re: _getPgGranuleByCollection.